File: backend/internal/service/notifications_27.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// NotificationRecord27Service encapsulates business rules for NotificationRecord27.
type NotificationRecord27Service struct {
	repo *repository.NotificationRecord27Repository
}

// NewNotificationRecord27Service wires the service.
func NewNotificationRecord27Service(repo *repository.NotificationRecord27Repository) *NotificationRecord27Service {
	return &NotificationRecord27Service{repo: repo}
}

// Create validates and persists a new NotificationRecord27.
func (s *NotificationRecord27Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.NotificationRecord27, error) {
	now := time.Now().UTC()
	rec := &models.NotificationRecord27{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a NotificationRecord27 by id.
func (s *NotificationRecord27Service) Get(ctx context.Context, id string) (*models.NotificationRecord27, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped NotificationRecord27 pages.
func (s *NotificationRecord27Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.NotificationRecord27, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing NotificationRecord27.
func (s *NotificationRecord27Service) Update(ctx context.Context, actor string, rec *models.NotificationRecord27) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a NotificationRecord27.
func (s *NotificationRecord27Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds NotificationRecord27 by external code prefix.
func (s *NotificationRecord27Service) Search(ctx context.Context, facilityID, prefix string) ([]models.NotificationRecord27, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *NotificationRecord27Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/notifications27.ts
/** Domain types for notifications / NotificationRecord27 */
export type NotificationRecord27Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface NotificationRecord27 {
  id: string;
  externalCode: string;
  displayName: string;
  status: NotificationRecord27Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface NotificationRecord27CreateRequest {
  externalCode: string;
  displayName: string;
  status: NotificationRecord27Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface NotificationRecord27ListResponse {
  items: NotificationRecord27[];
  total?: number;
}

export interface NotificationRecord27StatsResponse {
  activeCount: number;
}

export function isNotificationRecord27Active(rec: NotificationRecord27): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatNotificationRecord27Label(rec: NotificationRecord27): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareNotificationRecord27ByPriority(a: NotificationRecord27, b: NotificationRecord27): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
